api/src/main/java/edu/cornell/mannlib/vitro/webapp/application/VitroHomeDirectory.java

```diff
@@ -60,59 +71,209 @@ public String getDiscoveryMessage() {
-        log.info("Populating VIVO home at: " + destination.getPath());
-
-        TarArchiveEntry tarEntry;
-        try (TarArchiveInputStream tarInput = new TarArchiveInputStream(tar)) {
-            while ((tarEntry = tarInput.getNextTarEntry()) != null) {
-
-                // Use the example configurations
-                String outFilename = tarEntry.getName().replace("example.", "");
-                File outFile = new File(destination, outFilename);
-
-                // Is the entry a directory?
-                if (tarEntry.isDirectory()) {
-                    if (!outFile.exists()) {
-                        outFile.mkdirs();
-                    }
-
-                } else {
-                    // Entry is a File
-                    outFile.getParentFile().mkdirs();
-                    try (FileOutputStream fos = new FileOutputStream(outFile)) {
-                        IOUtils.copy(tarInput, fos);
-                    }
-                }
-            }
-        } catch (IOException e) {
-            throw new RuntimeException("Error creating home directory!", e);
-        }
-    }
+	/**
+	 * Populates VIVO home directory with files required to run.
+	 * 
+	 * NOTE: Will not overwrite any modified files on redeploy.
+	 */
+	public void populate() {
+		File vhdDir = getPath().toFile();
+
+		if (!vhdDir.isDirectory() || vhdDir.list() == null) {
+			throw new RuntimeException("Application home dir is not a directory! " + vhdDir);
+		}
+
+		Map<String, String> digest = untar(vhdDir);
+
+		writeDigest(digest);
+	}
+
+	/**
+	 * A non-destructive untar process that returns checksum digest of tarred files.
+	 * 
+	 * @param destination VIVO home directory
+	 * @return digest of each files checksum
+	 */
+	private Map<String, String> untar(File destination) {
+		log.info("Populating VIVO home at: " + destination.getPath());
+
+		Map<String, String> digest = new HashMap<>();
+		Map<String, String> storedDigest = loadDigest();
+
+		TarArchiveEntry tarEntry;
+		try (
+			InputStream homeDirTar = getHomeDirTar();
+			TarArchiveInputStream tarInput = new TarArchiveInputStream(homeDirTar);
+		) {
+			while ((tarEntry = tarInput.getNextTarEntry()) != null) {
+
+				// Use the example configurations
+				String outFilename = tarEntry.getName().replace("example.", "");
+				File outFile = new File(destination, outFilename);
+
+				// Is the entry a directory?
+				if (tarEntry.isDirectory()) {
+					if (!outFile.exists()) {
+						outFile.mkdirs();
+					}
+				} else {
+					// Entry is a File
+					boolean write = true;
+
+					// reading bytes into memory to avoid having to unreliable reset stream
+					byte[] bytes = IOUtils.toByteArray(tarInput);
+					digest.put(outFilename, checksum(bytes));
+
+					if (outFile.exists()) {
+						// if the stored digest contains the file and its checksum
+						// equals existing file's checksum, it is ok to overwrite
+						write = storedDigest.containsKey(outFilename)
```

Comment:
For existing installations, with existing files (starting from a release prior to this change), there is no digest file, and therefore, the tar entries are not written.
The question is, "how do we bootstrap the process of writing files from the most recent release into vivo-home"?
..or is it the responsibility of the sysadmin to delete the current files?

Reply:
Good catch. I had tried to be too clever and removed the `if (storedDigest.containsKey(outFilename))` before assigning the write variable. I intend for if no digest, it overwrites.